xmtp_mls/src/groups/device_sync.rs

```diff
@@ -101,57 +107,145 @@ pub enum DeviceSyncError {
     UnspecifiedDeviceSyncKind,
     #[error("sync reply is too old")]
     SyncReplyTimestamp,
+    #[error(transparent)]
+    Subscribe(#[from] SubscribeError),
+}
+
+impl<ApiClient, V> Client<ApiClient, V>
+where
+    ApiClient: XmtpApi + 'static,
+    V: SmartContractSignatureVerifier + 'static,
+{
+    pub async fn start_sync_worker(
+        &self,
+        provider: &XmtpOpenMlsProvider,
+    ) -> Result<(), DeviceSyncError> {
+        self.sync_init(provider).await?;
+
+        crate::spawn(None, {
+            let client = self.clone();
+
+            let receiver = client.local_events.subscribe();
+            let mut sync_stream = Box::pin(receiver.stream_sync_messages());
+
+            async move {
+                while let Err(err) = client.sync_worker(&mut sync_stream).await {
+                    tracing::error!("Sync worker error: {err}");
+                }
+            }
+        });
+
+        Ok(())
+    }
 }
 
 impl<ApiClient, V> Client<ApiClient, V>
 where
     ApiClient: XmtpApi,
     V: SmartContractSignatureVerifier,
 {
-    pub async fn reply_to_sync_request(
+    pub(crate) async fn sync_worker(
         &self,
-        provider: &XmtpOpenMlsProvider,
-        kind: DeviceSyncKind,
-    ) -> Result<DeviceSyncReplyProto, DeviceSyncError> {
-        let conn = provider.conn_ref();
+        sync_stream: &mut Pin<Box<impl Stream<Item = Result<SyncMessage, SubscribeError>>>>,
+    ) -> Result<(), DeviceSyncError> {
+        let mut sync_stream = sync_stream.as_mut();
 
-        let (_msg, request) = self.pending_sync_request(provider, kind).await?;
-        let records = match kind {
-            DeviceSyncKind::Consent => vec![self.syncable_consent_records(conn)?],
-            DeviceSyncKind::MessageHistory => {
-                vec![self.syncable_groups(conn)?, self.syncable_messages(conn)?]
+        let provider = self.mls_provider()?;
+
+        let get_message_debounce = |conn: &DbConnection, message_id: &[u8]| {
+            let mut tries = 0;
+            while tries < 3 {
```

Comment:
Could also use `crate::retry::retry_async!` macro here: [macro](https://xmtp.github.io/libxmtp/xmtp_mls/macro.retry_async.html) and optional [builder](https://xmtp.github.io/libxmtp/xmtp_mls/retry/index.html)